presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/ReorderJoinsUtils.java

```diff
@@ -0,0 +1,183 @@
+            }
+        }
+
+        ImmutableList.Builder<JoinNode.EquiJoinClause> leftCriteria = ImmutableList.builder();
+        ImmutableList.Builder<JoinNode.EquiJoinClause> rightCriteria = ImmutableList.builder();
+        ImmutableList.Builder<JoinNode.EquiJoinClause> spanningCriteria = ImmutableList.builder();
+        for (JoinNode.EquiJoinClause equiJoinClause : joinGraph.getCriteria()) {
+            Set<Symbol> equiJoinClauseSymbols = ImmutableSet.of(equiJoinClause.getLeft(), equiJoinClause.getRight());
+            if (leftSymbols.containsAll(equiJoinClauseSymbols)) {
+                leftCriteria.add(equiJoinClause);
+            }
+            else if (rightSymbols.containsAll(equiJoinClauseSymbols)) {
+                rightCriteria.add(equiJoinClause);
+            }
+            else {
+                spanningCriteria.add(equiJoinClause);
+            }
+        }
+
+        ImmutableList.Builder<Expression> leftFilters = ImmutableList.builder();
+        ImmutableList.Builder<Expression> rightFilters = ImmutableList.builder();
+        ImmutableList.Builder<Expression> spanningFilters = ImmutableList.builder();
+        for (Expression filter : joinGraph.getFilters()) {
+            Set<Symbol> filterSymbols = SymbolReferenceExtractor.getSymbolReferences(filter).stream()
+                    .map(Symbol::from)
+                    .collect(toImmutableSet());
+            if (leftSymbols.containsAll(filterSymbols)) {
+                leftFilters.add(filter);
+            }
+            else if (rightSymbols.containsAll(filterSymbols)) {
+                rightFilters.add(filter);
+            }
+            else {
+                spanningFilters.add(filter);
+            }
+        }
+
+        PlanNode left = getJoinSource(idAllocator, leftNodes.build(), leftFilters.build(), leftCriteria.build());
+        PlanNode right = getJoinSource(idAllocator, rightNodes.build(), rightFilters.build(), rightCriteria.build());
+
+        List<Expression> filters = spanningFilters.build();
+
+        return new JoinNode(
+                idAllocator.getNextId(),
+                JoinNode.Type.INNER,
+                left,
+                right,
+                flipJoinCriteriaIfNecessary(spanningCriteria.build(), left),
+                ImmutableList.<Symbol>builder().addAll(left.getOutputSymbols()).addAll(right.getOutputSymbols()).build(),
+                filters.isEmpty() ? Optional.empty() : Optional.of(ExpressionUtils.and(filters)),
+                Optional.empty(),
+                Optional.empty(),
+                Optional.of(PARTITIONED));
+    }
+
+    private static List<JoinNode.EquiJoinClause> flipJoinCriteriaIfNecessary(List<JoinNode.EquiJoinClause> joinCriteria, PlanNode left)
+    {
+        return joinCriteria
+                .stream()
+                .map(criterion -> left.getOutputSymbols().contains(criterion.getLeft()) ? criterion : new JoinNode.EquiJoinClause(criterion.getRight(), criterion.getLeft()))
```

Comment:
You could define `JoinNode.EquiJoinClause.flip()` and use it here and in `JoinNode.flip`